File: python/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#


# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        pairs = {
            ']' : '[',
            '}' : '{',
            ')' : '(',
        }
        stack = list()
        for ch in s:
            if ch in pairs:
                if not stack or stack[-1] != pairs[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)
        return not stack

    # 丑陋的代码
    def isValid1(self, s: str) -> bool:
        stack = list()
        for _, ch in enumerate(s):
            if ch == '(' or ch == '[' or ch == '{':
                stack.append(ch)
            elif ch == ')':
                if not stack:
                    return False
                out = stack.pop()
                if out != '(':
                    return False
            elif ch == '}':
                if not stack:
                    return False
                out = stack.pop()
                if out != '{':
                    return False
            elif ch == ']':
                if not stack:
                    return False
                out = stack.pop()
                if out != '[':
                    return False
            else:
                return False
        if stack:
            return False
        return True


# Solution().isValid('()')

# @lc code=end

File: cpp/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    /**
     * 栈 + 表
     */
    bool isValid(string s)
    {
        if (s.size() % 2 != 0) {
            return false;
        }

        unordered_map<char, char> table{{'(', ')'}, {'[', ']'}, {'{', '}'}};

        vector<char> m_stack;
        for (char ch : s) {
            if (ch == '(' || ch == '[' || ch == '{') {
                m_stack.emplace_back(ch);
            }
            else {
                if (m_stack.empty()) {
                    return false;
                }
                char top = m_stack.back();
                m_stack.pop_back();
                if (table[top] != ch) {
                    return false;
                }
            }
        }
        return m_stack.empty();
    }
};
// @lc code=end
